zkvm/src/predicate.rs

```diff
@@ -136,43 +102,61 @@ impl Predicate {
         Predicate::Opaque(self.to_point())
     }
 
-    /// Creates a disjunction of a vector of predicates.
-    pub fn disjunction(preds: Vec<Predicate>) -> Result<Self, VMError> {
+    /// Creates a Tree from a vector of programs.
+    pub fn tree(key: Option<VerificationKey>, progs: Vec<Program>, blinding_key: [u8; 32]) -> Result<Self, VMError> {
+        let signing_key = match key {
+            Some(x) => x,
+            _ => VerificationKey(PedersenGens::default().B_blinding.compress()),
+        };
+        // PRTODO: Do I assume that progs already contains the per-program blinding factors?
```

Comment:
Not quite. The idea is to use some tree leaves as programs, and others - as blinding factors for them. Meaning, each leaf-program is replaced by a node with two leaves: a program and a dummy leaf acting as additional entropy to blind the program.

When constructing a tree, a single `blinding_key` can be used to generate blinding factors for each program. This can be done by creating a new transcript instance with label, say, `"ZkVM.taproot-blinding"`, commit N programs, and squeeze N blinding factors. Then, the first bit of the blinding factor can be used to determine the relative position of the program and a blinding factor: 0 for blinding-on-the-left, 1 for blinding-on-the-right.

```
N=4
progs = [p1, p2, p3, p4]
T = Transcript::new("ZkVM.taproot-blinding")
T.commit("n", 4)
T.commit("prog", p1)
T.commit("prog", p2)
T.commit("prog", p3)
T.commit("prog", p4)

f1 = T.challenge("blinding") // 1...
f2 = T.challenge("blinding") // 0...
f3 = T.challenge("blinding") // 0...
f4= T.challenge("blinding") // 1...

leaves = [p1, f1, f2, p2, f3, p3, p4, f4]
root = MerkleTree::root(leaves)
```

Reply:
The `zkvm::Verifier` will not know the difference between a neighbour being a hash of a blinding factor, and a neighbour being a hash of program. Also, the relative order is uniformly random, so even the fact of using this particular scheme remains hidden.

Reply:
The last piece: in multi-party contracts, Alice wants to be sure that Bob cannot use their blinding factor as a valid program (and vice versa). So we'd use a different leaf hashing function to make the resulting merkle node non-openable into a program.

Currently, `impl MerkleItem for Program` uses label `"program"` for committing a program. For blinding factor, we need a domain separator, like `"blinding"`. Then, Bob can show the blinding factor to Alice and she'll compute the node with a different label, and will be convinced that it's infeasible to find a program that (if committed with a proper `"program"` label) would open to the same node hash.